File: apps/mobile-ios/Sources/App/APIClient.swift
import Foundation

// MARK: - API Errors

enum APIError: LocalizedError {
    case invalidURL
    case invalidResponse
    case httpError(statusCode: Int, apiError: APIErrorResponse?)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .invalidResponse:
            return "Invalid server response"
        case .httpError(let code, let apiError):
            return apiError?.message ?? "HTTP error \(code)"
        }
    }
}

// MARK: - API Client

final class APIClient {
    let baseURL: String
    var sessionToken: String?

    init(baseURL: String = AppConfig.apiBaseURL.absoluteString) {
        self.baseURL = baseURL
    }

    // MARK: Core request

    private func performRequest<T: Decodable>(
        _ path: String,
        method: String = "GET",
        bodyData: Data? = nil
    ) async throws -> T {
        guard let url = URL(string: baseURL + path) else {
            throw APIError.invalidURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(UUID().uuidString, forHTTPHeaderField: "X-Correlation-ID")
        if let token = sessionToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = bodyData

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let http = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }
        guard (200...299).contains(http.statusCode) else {
            let apiErr = try? JSONDecoder().decode(APIErrorResponse.self, from: data)
            throw APIError.httpError(statusCode: http.statusCode, apiError: apiErr)
        }

        return try JSONDecoder().decode(T.self, from: data)
    }

    // MARK: Auth (unauthenticated)

    func pkceInit() async throws -> APIResponse<PKCEInitResponse> {
        try await performRequest("/v1/auth/session/pkce-init", method: "POST")
    }

    func exchangeCode(_ code: String, state: String) async throws -> APIResponse<ExchangeResponse> {
        struct Body: Encodable { let code: String; let state: String }
        let data = try JSONEncoder().encode(Body(code: code, state: state))
        return try await performRequest("/v1/auth/session/exchange", method: "POST", bodyData: data)
    }

    // MARK: User

    func getMe() async throws -> APIResponse<UserResponse> {
        try await performRequest("/v1/me")
    }

    // MARK: Projects

    func listProjects(cursor: String? = nil) async throws -> PaginatedResponse<Project> {
        var path = "/v1/projects"
        if let cursor { path += "?cursor=\(cursor)" }
        return try await performRequest(path)
    }

    func getProject(_ id: String) async throws -> APIResponse<Project> {
        try await performRequest("/v1/projects/\(id)")
    }

    // MARK: Sessions (Runs)

    func listSessions(projectId: String, cursor: String? = nil) async throws -> PaginatedResponse<Session> {
        var path = "/v1/projects/\(projectId)/sessions"
        if let cursor { path += "?cursor=\(cursor)" }
        return try await performRequest(path)
    }

    func getSession(_ id: String) async throws -> APIResponse<Session> {
        try await performRequest("/v1/sessions/\(id)")
    }

    func startSession(_ id: String) async throws -> APIResponse<Session> {
        try await performRequest("/v1/sessions/\(id)/start", method: "POST")
    }

    func stopSession(_ id: String) async throws -> APIResponse<Session> {
        try await performRequest("/v1/sessions/\(id)/stop", method: "POST")
    }

    // MARK: Comments

    func listComments(projectId: String, cursor: String? = nil) async throws -> PaginatedResponse<CollaborationThread> {
        var path = "/v1/projects/\(projectId)/comments"
        if let cursor { path += "?cursor=\(cursor)" }
        return try await performRequest(path)
    }

    func postComment(projectId: String, commentBody: String, threadId: String? = nil) async throws -> APIResponse<Comment> {
        struct Anchor: Encodable { let type: String }
        struct Payload: Encodable { let body: String; let threadId: String?; let anchor: Anchor? }
        let payload = Payload(body: commentBody, threadId: threadId, anchor: threadId == nil ? Anchor(type: "general") : nil)
        let data = try JSONEncoder().encode(payload)
        return try await performRequest("/v1/projects/\(projectId)/comments", method: "POST", bodyData: data)
    }

    // MARK: Previews

    func createPreview(sessionId: String) async throws -> APIResponse<PreviewRouteBinding> {
        try await performRequest("/v1/sessions/\(sessionId)/previews", method: "POST")
    }

    func getPreview(_ previewId: String) async throws -> APIResponse<PreviewRouteBinding> {
        try await performRequest("/v1/previews/\(previewId)")
    }

    func getPreviewToken(previewId: String) async throws -> APIResponse<PreviewTokenResponse> {
        try await performRequest("/v1/previews/\(previewId)/token", method: "POST")
    }
}

File: apps/web/src/lib/api-client.ts
// ============================================================
// Typed API client — wraps all platform API calls
// Phase 2: connect to real endpoints
// ============================================================

import type {
  ExchangeTokenRequest, ExchangeTokenResponse,
  MeResponse, CreateWorkspaceRequest, CreateProjectRequest,
  CreateSessionRequest, CreatePreviewRequest,
  CreateProviderConfigRequest, UpdateProviderConfigRequest,
  RotateSecretRequest, CreateAgentRoleRequest,
  CreatePipelineRequest, CreatePipelineRunRequest,
  ApiResponse, PaginatedResponse,
} from '@udd/contracts';
import type { Workspace, Project, Session, PreviewRouteBinding } from '@udd/contracts';
import type { ProviderConfig, AgentRole, PipelineDefinition, PipelineRunRecord } from '@udd/contracts';

const API_BASE_URL = process.env['NEXT_PUBLIC_API_URL'] ?? '/v1';

export class ApiClient {
  private readonly baseUrl: string;
  private sessionToken: string | null = null;

  constructor(baseUrl = API_BASE_URL) {
    this.baseUrl = baseUrl;
  }

  setSessionToken(token: string): void {
    this.sessionToken = token;
  }

  private async request<T>(
    method: string,
    path: string,
    body?: unknown,
    options?: { correlationId?: string },
  ): Promise<T> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    if (this.sessionToken) {
      headers['Authorization'] = `Bearer ${this.sessionToken}`;
    }
    if (options?.correlationId) {
      headers['x-correlation-id'] = options.correlationId;
    }

    const res = await fetch(`${this.baseUrl}${path}`, {
      method,
      headers,
      body: body !== undefined ? JSON.stringify(body) : null,
    });

    const json = await res.json() as T;

    if (!res.ok) {
      throw Object.assign(new Error((json as { message?: string }).message ?? 'API error'), {
        status: res.status,
        body: json,
      });
    }

    return json;
  }

  // Auth
  exchangeToken(req: ExchangeTokenRequest): Promise<ExchangeTokenResponse> {
    return this.request('POST', '/auth/session/exchange', req);
  }

  getMe(): Promise<ApiResponse<MeResponse>> {
    return this.request('GET', '/me');
  }

  // Workspaces
  listWorkspaces(): Promise<PaginatedResponse<Workspace>> {
    return this.request('GET', '/workspaces');
  }

  createWorkspace(req: CreateWorkspaceRequest): Promise<ApiResponse<Workspace>> {
    return this.request('POST', '/workspaces', req);
  }

  getWorkspace(id: string): Promise<ApiResponse<Workspace>> {
    return this.request('GET', `/workspaces/${id}`);
  }

  // Projects
  listProjects(workspaceId: string): Promise<PaginatedResponse<Project>> {
    return this.request('GET', `/workspaces/${workspaceId}/projects`);
  }

  createProject(workspaceId: string, req: CreateProjectRequest): Promise<ApiResponse<Project>> {
    return this.request('POST', `/workspaces/${workspaceId}/projects`, req);
  }

  // Sessions
  createSession(projectId: string, req?: CreateSessionRequest): Promise<ApiResponse<Session>> {
    return this.request('POST', `/projects/${projectId}/sessions`, req ?? {});
  }

  startSession(sessionId: string): Promise<ApiResponse<Session>> {
    return this.request('POST', `/sessions/${sessionId}/start`, {});
  }

  stopSession(sessionId: string): Promise<ApiResponse<Session>> {
    return this.request('POST', `/sessions/${sessionId}/stop`, {});
  }

  // Previews
  createPreview(sessionId: string, req: CreatePreviewRequest): Promise<ApiResponse<PreviewRouteBinding>> {
    return this.request('POST', `/sessions/${sessionId}/previews`, req);
  }

  // AI Providers
  listProviders(): Promise<PaginatedResponse<ProviderConfig>> {
    return this.request('GET', `/me/ai/providers`);
  }

  createProvider(req: CreateProviderConfigRequest): Promise<ApiResponse<ProviderConfig>> {
    return this.request('POST', `/me/ai/providers`, req);
  }

  updateProvider(providerId: string, req: UpdateProviderConfigRequest): Promise<ApiResponse<ProviderConfig>> {
    return this.request('PATCH', `/me/ai/providers/${providerId}`, req);
  }

  rotateProviderSecret(providerId: string, req: RotateSecretRequest): Promise<ApiResponse<void>> {
    return this.request('POST', `/me/ai/providers/${providerId}/rotate-secret`, req);
  }

  deleteProvider(providerId: string): Promise<ApiResponse<void>> {
    return this.request('DELETE', `/me/ai/providers/${providerId}`);
  }

  // AI Roles
  listRoles(projectId: string): Promise<PaginatedResponse<AgentRole>> {
    return this.request('GET', `/projects/${projectId}/ai/roles`);
  }

  createRole(projectId: string, req: CreateAgentRoleRequest): Promise<ApiResponse<AgentRole>> {
    return this.request('POST', `/projects/${projectId}/ai/roles`, req);
  }

  // AI Pipelines
  listPipelines(projectId: string): Promise<PaginatedResponse<PipelineDefinition>> {
    return this.request('GET', `/projects/${projectId}/ai/pipelines`);
  }

  createPipeline(projectId: string, req: CreatePipelineRequest): Promise<ApiResponse<PipelineDefinition>> {
    return this.request('POST', `/projects/${projectId}/ai/pipelines`, req);
  }

  // Pipeline Runs
  listRuns(projectId: string): Promise<PaginatedResponse<PipelineRunRecord>> {
    return this.request('GET', `/projects/${projectId}/ai/runs`);
  }

  createRun(projectId: string, pipelineId: string, req: CreatePipelineRunRequest): Promise<ApiResponse<PipelineRunRecord>> {
    return this.request('POST', `/projects/${projectId}/ai/pipelines/${pipelineId}/runs`, req);
  }

  cancelRun(projectId: string, runId: string): Promise<ApiResponse<PipelineRunRecord>> {
    return this.request('POST', `/projects/${projectId}/ai/runs/${runId}/cancel`, {});
  }

  // PKCE init — returns state nonce and code challenge for WorkOS auth redirect
  async pkceInit(): Promise<{ data: { state: string; codeChallenge: string; codeChallengeMethod: string } }> {
    return this.request('POST', '/auth/session/pkce-init');
  }

  // Exchange authorization code + PKCE state for a session JWT
  async exchangeCode(code: string, state: string): Promise<{ data: { token: string; userId: string } }> {
    return this.request('POST', '/auth/session/exchange', { code, state });
  }
}

// Singleton for client-side usage
export const apiClient = new ApiClient();
